### PR TITLE
Allows usage of `URL` pointing to a local file in file operations

### PR DESCRIPTION
**Pull request by Aschen**
## Description

Allows to provide an `URL` object pointing to a local file with `CommitFile` operation.

Providing a non-local file will throw an error.

This is not available in the browser.

Fix https://github.com/huggingface/huggingface.js/issues/106

```
await commit({
  repo,
  title: "Some commit",
  credentials: {
    accessToken: TEST_ACCESS_TOKEN,
  },
  operations: [
    {
      operation: "addOrUpdate",
      content: new URL("file://tsconfig.json"),
      path: "tsconfig.json",
    },
  ],
});
```

### CODE REVIEW

File: packages/hub/src/lib/commit.ts [hunk 1]
```diff
@@ -78,6 +78,49 @@ function isFileOperation(op: CommitOperation): op is CommitFile {
 	return op.operation === "addOrUpdate";
 }
 
+const isBrowser = typeof window !== "undefined" && typeof window.document !== "undefined";
```
Comment (coyotte508) [inline packages/hub/src/lib/commit.ts:81]: Let's reuse the full check from `sha256.ts`:

```ts
	const isBrowser = typeof window !== "undefined" && typeof window.document !== "undefined";

	const isWebWorker =
		typeof self === "object" && self.constructor && self.constructor.name === "DedicatedWorkerGlobalScope";
```

File: packages/hub/src/lib/commit.ts [hunk 2]
```diff
@@ -78,6 +78,49 @@ function isFileOperation(op: CommitOperation): op is CommitFile {
...
+		const { LazyBlob } = isBrowser ? { LazyBlob: FakeLazyBlob } : await import("../utils/LazyBlob");
+
+		const blobMap = new FileBlobs();
+
+		for (const operation of operations) {
+			if (operation.content instanceof Blob) {
+				blobMap.set(operation.content, operation.content);
+			} else {
+				if (operation.content.protocol !== "file:") {
+					throw TypeError('Only "file://" protocol is supported for now');
```
Comment (coyotte508) [inline packages/hub/src/lib/commit.ts:107]:
```suggestion
					throw TypeError('Only "file://" protocol is supported for URLs');
```

File: packages/hub/src/lib/commit.ts [hunk 3]
```diff
@@ -89,18 +132,20 @@ async function* commitIter(params: CommitParams): AsyncGenerator<unknown, Commit
 
 	const lfsShas = new Map<string, string | null>();
 
+	const blobs = await FileBlobs.createBlobMap(params.operations.filter(isFileOperation));
```
Comment (coyotte508) [inline packages/hub/src/lib/commit.ts:135]:
```suggestion
	// Convert `operation.content` to `Blob`
	const operations = await Promise.all(params.operations.map(op => op.operation === "addOrUpdate" ? {...op, content: await ...} : op));
	
	...
```

No need to create/override a separate `Map` class

Comment (Aschen) [reply packages/hub/src/lib/commit.ts:135]: I did it to avoid casting systematically the `content` to `Blob` but I realize that's a very poor design choice :roll_eyes:

Comment (coyotte508) [reply packages/hub/src/lib/commit.ts:135]: (and changing `isFileOperation` too)

Comment (Aschen) [reply packages/hub/src/lib/commit.ts:135]: I tried this but unfortunately since `operation.content` is a `ContentSource`, ts complain that it could be `URL` and not `Blob`.

```js
const operations = await Promise.all(
  params.operations.map(async (operation) => {
    if (isFileOperation(operation) && operation.content instanceof URL) {
      if (operation.content.protocol !== "file:") {
        throw TypeError('Only "file://" protocol is supported for now');
      }

      // Ignore the "file://" at the beginning and the trailing slash
      const lazyBlob = await LazyBlob.create(operation.content.href.slice(7, operation.content.href.length - 1));

      return {
        ...operation,
        content: lazyBlob,
      };
    }

    return operation;
  })
);
```

**Comment by Aschen** [comment]
@coyotte508 About handling file from the browser side, shouldn't we support [File](https://w3c.github.io/FileAPI/) object that can be get when using the file picker API? 

See also https://developer.chrome.com/articles/file-system-access/

File: packages/hub/src/lib/commit.ts [hunk 4]
```diff
@@ -196,9 +235,9 @@ async function* commitIter(params: CommitParams): AsyncGenerator<unknown, Commit
 				}
 
 				if (obj.error) {
-					const errorMessage = `Error while doing LFS batch call for ${operations[shas.indexOf(obj.oid)].path}: ${
-						obj.error.message
-					}${batchRequestId ? ` - Request ID: ${batchRequestId}` : ""}`;
+					const errorMessage = `Error AsyncGeneratorError while doing LFS batch call for ${
+						operations[shas.indexOf(obj.oid)].path
+					}: ${obj.error.message}${batchRequestId ? ` - Request ID: ${batchRequestId}` : ""}`;
```
Comment (coyotte508) [inline packages/hub/src/lib/commit.ts:240]: Why the change?

File: packages/hub/src/lib/commit.ts [hunk 5]
```diff
@@ -89,11 +99,40 @@ async function* commitIter(params: CommitParams): AsyncGenerator<unknown, Commit
 
 	const lfsShas = new Map<string, string | null>();
 
-	const gitAttributes = (
-		params.operations.find((op) => isFileOperation(op) && op.path === ".gitattributes") as CommitFile | undefined
-	)?.content;
+	const allOperations = await Promise.all(
+		params.operations.map(async (operation) => {
+			if (operation.operation === "addOrUpdate") {
```
Comment (Aschen) [inline packages/hub/src/lib/commit.ts:104]: Actually here (and l.105 on the second if), you could use a guard to save 2 levels of indentation

**Review by coyotte508 — APPROVED** [review]
Nice work!
